### PR TITLE
Newton solver extensions

### LINKED ISSUE #842: Add Newton solver extensions
Merge the Newton solver improvements from

G. T. Lines, @paszkow, L. Schmiester, D. Weindl, P. Stapor, and J. Hasenauer. Efficient computation of steady states in large-scale ODE models of biochemical reaction networks. accepted for Proceedings of the 8th IFAC Conference on Foundations of Systems Biology in Engineering (FOSBE), Valencia, Spain, October 2019.

### PR DESCRIPTION
**Pull request by paszkow**
The changes that allow performing Newton tests from the paper:

G. T. Lines, Ł. Paszkowski, L. Schmiester, D. Weindl, P. Stapor, and J. Hasenauer. Efficient computation of steady states in large-scale ODE models of biochemical reaction networks. accepted for Proceedings of the 8th IFAC Conference on Foundations of Systems Biology in Engineering (FOSBE), Valencia, Spain, October 2019.

Closes #842

### CODE REVIEW

**Comment by dweindl** [comment]
Thanks @paszkow . @paulstapor , can you please look into that when time permits?

**Comment by paszkow** [comment]
@dweindl @paulstapor You might want to hold on for a while. Turn out I cannot reproduce all of the tests from the paper. There might be something missing here.

**Comment by paulstapor** [comment]
Thanks for your work @paszkow , and thanks for your message, good to know.
We are also busy with fixing #849, just poke me as soon as you think I should do a code review.

File: include/amici/defines.h [hunk 1]
```diff
@@ -148,6 +148,12 @@ enum class NewtonStatus {
     newt_sim_newt=3,
 };
 
+/** Damping factor flag for the Newton method */
+enum class NewtonDampingFactorMode {
+    off = 0,
+    on = 1
```
Comment (paulstapor) [inline include/amici/defines.h]: Just to make sure: Default should then be `on`, right?

File: include/amici/newton_solver.h [hunk 2]
```diff
@@ -103,6 +105,8 @@ class NewtonSolver {
     double atol = 1e-16;
     /** relative tolerance */
     double rtol = 1e-8;
+    /** damping factor flag */
+    NewtonDampingFactorMode dampingFactorMode = NewtonDampingFactorMode::on;
```
Comment (paulstapor) [inline include/amici/newton_solver.h]: There we go, yes... Ignore the comment above...

**Comment by paszkow** [comment]
I think I have pushed all relevant changes. After a proper re-generating the model and tweaking the test code I can reproduce the simulations.

File: src/hdf5.cpp [hunk 3]
```diff
@@ -628,6 +628,12 @@ void readSolverSettingsFromHDF5(H5::H5File const& file, Solver &solver,
                     getIntScalarAttribute(file, datasetPath, "newton_preeq"));
     }
 
+    if(attributeExists(file, datasetPath, "newton_damping_factor_mode")) {
+        solver.setNewtonDampingFactorMode(
+                    static_cast<NewtonDampingFactorMode>(
+                        getIntScalarAttribute(file, datasetPath, "newton_damping_factor_mode")));
+    }
+
```
Comment (paulstapor) [inline src/hdf5.cpp]: Yep, makes sense that this had an impact on reproducing the results... ;)

File: src/steadystateproblem.cpp [hunk 4]
```diff
@@ -206,21 +207,16 @@ void SteadystateProblem::applyNewtonsMethod(ReturnData *rdata, Model *model,
...
             converged = wrms < RCONST(1.0);
 
-            if (converged) {
-                /* Ensure positivity of the found state */
-                for (ix = 0; ix < model->nx_solver; ix++) {
-                    if (x[ix] < 0.0) {
-                        x[ix] = 0.0;
-                        converged = FALSE;
-                    }
-                }
```
Comment (paulstapor) [inline src/steadystateproblem.cpp]: What is the reason to kick out this part? Usually, we want positive states...
Sure, this part may screw up convergence, if we converged to a steady state with negative entries... But usually, we don't want to keep this steady state with negative entries anyway...

File: src/steadystateproblem.cpp [hunk 5]
```diff
@@ -206,21 +207,16 @@ void SteadystateProblem::applyNewtonsMethod(ReturnData *rdata, Model *model,
...
+            if (!converged && newtonSolver->dampingFactorMode==NewtonDampingFactorMode::on) {
+              /* increase dampening factor (superfluous, if converged) */
+              gamma = fmin(1.0, 2.0 * gamma);
             }
-        } else {
-            /* Reduce dampening factor */
+        } else if (newtonSolver->dampingFactorMode==NewtonDampingFactorMode::on) {
+            /* Reduce dampening factor and raise an error when becomes too small */
             gamma = gamma / 4.0;
+            if (gamma < 1e-6)
```
Comment (paulstapor) [inline src/steadystateproblem.cpp]: How sure are we about this number?
Might it be good to make this an user-adjustable parameter?

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paszkow) [reply src/steadystateproblem.cpp]: Makes problems when you have a zero steady-state (or at least zero at one index). Then you will never converge due to floating-point operations. The number -1e-32 is so close to zero that you do not want to force it to be 0.0. We had such issues.

On the other hand, for strictly positive steady-states this part of the code was never invoked.

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (paszkow) [reply src/steadystateproblem.cpp]: Up to you. It is just a reasonable lower bound. Usually, you do not need to go lower. If you need, then most likely you will not converge at all.

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: I see your point... but:
This part of Amici is sometimes used to preequilibrate systems and start a simulation after prequilibration, i.e., sometimes, the Ewton-solver is used to find inital values of ODEs. Negative states can then cause integration failure, when a simulation is started from them... We also experienced cases like this...

So, we must find a solution to this... 
I don't yet really understand what goes wrong if we just set those `x` to `0.0`. Can you re-explain, please?

About strict positivity: At least to my understanding, this is not what one is looking for. It is important to be `>= 0.0`. Please correct me here, if I'm wrong. Maybe @FFroehlich knows more about these cases than I do...

How about checking whether we are in between of `-absolute_tolerance` and `0.0`, an just flip sign in that case?

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: I see the point, but I like avoiding magic numbers which are hard-coded in algorithms...
As the damping factor is something like a step size, I'm wondering whether linking it to the error tolerances might be a good idea...

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paszkow) [reply src/steadystateproblem.cpp]: I see the issues. Maybe a reasonable solution would be to have `if (x[i] < -eps)` for some epsilon.

In our case, with this part of the code and zero steady-states, the Newton solver never converged. Always reached the maximum number of iterations. In the debugging, I noticed the solution was forced to be zero and in the next step, it again went to negative (-1e-32). The process repeated all the time.

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (dweindl) [reply src/steadystateproblem.cpp]: Agreed with the magic numbers. Maybe for start just add them as function arguments with the given default values. Linking to error tolerances seems like a start of a new project. Sounds good, but maybe too much here.

Comment (paszkow) [reply src/steadystateproblem.cpp]: Totally up to you.

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: Ah, okay, now I see what's wrong:
You're totally right about taking out the `converged = FALSE;`, indeed!
One thing: Can we agree on setting stuff to 0.0, and then just rechecking for convergence?
If convergence persists, we are fine and keep `converged = TRUE;` and hence drop out of the loop. Only if we have no convergence, we set  `converged = FALSE;` after reapplying the positivity constraint?

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: Let's make it a user option for the moment.

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paszkow) [reply src/steadystateproblem.cpp]: Understand. I will check it out. Could be enough indeed.

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (paszkow) [reply src/steadystateproblem.cpp]: OK

Comment (paszkow) [reply src/steadystateproblem.cpp]: Is 1e-6 a reasonable default value or you prefer a different number?

Comment (paulstapor) [reply src/steadystateproblem.cpp]: Seems reasonable to me. Either that or 1e-8, which is our default for relative tolerances... Both are fine for me.

File: src/steadystateproblem.cpp (hunk 4 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: Does your code example still work with this adaptation now? If yes, I'd do a review for merging.

Comment (paszkow) [reply src/steadystateproblem.cpp]: I wouldn't have done this change without running my test first. It works indeed.

Comment (paulstapor) [reply src/steadystateproblem.cpp]: Okay, point for you. ;)

**Comment by paulstapor** [comment]
Little error persisting on the documentation:
```
DOXYGEN failed:
3559/Users/travis/build/ICB-DCM/AMICI/include/amici/solver.h:228: warning: argument 'damping' of command @param is not found in the argument list of Solver::setNewtonDampingFactorMode(NewtonDampingFactorMode dampingFactorMode)
3560/Users/travis/build/ICB-DCM/AMICI/include/amici/solver.h:232: warning: The following parameters of amici::Solver::setNewtonDampingFactorMode(NewtonDampingFactorMode dampingFactorMode) are not documented:
3561  parameter 'dampingFactorMode'
3562/Users/travis/build/ICB-DCM/AMICI/include/amici/solver.h:240: warning: argument 'damping' of command @param is not found in the argument list of Solver::setNewtonDampingFactorLowerBound(double dampingFactorLowerBound)
3563/Users/travis/build/ICB-DCM/AMICI/include/amici/solver.h:244: warning: The following parameters of amici::Solver::setNewtonDampingFactorLowerBound(double dampingFactorLowerBound) are not documented:
3564  parameter 'dampingFactorLowerBound'
```

**Comment by dweindl** [comment]
CI failure is due to an unrelated doxygen problem. Consider that passed.

**Comment by dweindl** [comment]
@paszkow : Can you please add `Model::newton_damping_factor_mode` and `Model::newton_damping_factor_lower_bound` to the respective function in `include/amici/serialization.h`, then we are done, I think.

**Comment by paszkow** [comment]
Sorry. I missed the serialization part.

**Pull request merged**